addons/mail/models/mail_scheduled_message.py

```diff
@@ -0,0 +1,262 @@
+            if raise_exception is set to False, the method will skip the posting of a message
+            instead of raising an error. This is useful when scheduled messages are sent from
+            the _post_messages_cron
+        """
+        notification_parameters_whitelist = self._notification_parameters_whitelist()
+        for scheduled_message in self:
+            message_creator = scheduled_message.create_uid
+            try:
+                scheduled_message.with_user(message_creator)._check()
+            except AccessError:
+                if raise_exception:
+                    raise
+                _logger.info("Posting of scheduled message %s failed: user %s cannot post on the record", scheduled_message.id, message_creator.id)
+                continue
+            self.env[scheduled_message.model].browse(scheduled_message.res_id).with_user(message_creator).message_post(
+                attachment_ids=list(scheduled_message.attachment_ids.ids),
+                author_id=scheduled_message.author_id.id,
+                body=scheduled_message.body,
+                partner_ids=list(scheduled_message.partner_ids.ids),
+                subtype_xmlid='mail.mt_note' if scheduled_message.is_note else 'mail.mt_comment',
+                **{k: v for k, v in json.loads(scheduled_message.notification_parameters or '{}').items() if k in notification_parameters_whitelist},
+            )
+        self.unlink()
+
+    # ------------------------------------------------------
+    # Business Methods
+    # ------------------------------------------------------
+
+    @api.model
+    def _check(self, values=None):
+        """ Restrict the access to a scheduled message.
+            Access is based on the record on which the scheduled message will be posted to
+        """
+        if self.env.is_superuser():
+            return True
+
+        model_ids = defaultdict(set)
+        if self:
+            for scheduled_message in self:
+                model_ids[scheduled_message.model].add(scheduled_message.res_id)
+        if values and (model := values.get('model')) and (res_id := values.get('res_id')):
+            model_ids[model].add(res_id)
+
+        for model, res_ids in model_ids.items():
+            records = self.env[model].browse(res_ids)
+            operation = records._mail_post_access
+            records.check_access(operation)
+
+    @api.model
+    def _notification_parameters_whitelist(self):
+        """ Parameters that can be used when posting the scheduled messages.
+        """
+        return {
+            'email_add_signature',
+            'email_from',
+            'email_layout_xmlid',
+            'force_email_lang',
+            'mail_activity_type_id',
+            'mail_auto_delete',
+            'mail_server_id',
+            'message_type',
+            'model_description',
+            'parent_id',
+            'record_alias_domain_id',
+            'record_company_id',
```

Comment:
Aren't those computed during message_post ?

Reply:
To be honest I just took the ones returned by the composer's `_prepare_mail_values` method.
These ones might indeed be computed in message_post but we can pass them as params.
Should we remove them to make sure they are "up to date" at the time the messages are posted (currently done)? Or does it not matter and we can keep them to not recompute them maybe